### PR TITLE
feat: Add support for fetching Audit Records

### PR DESCRIPTION
**Pull request by t-junjie**
See issue [394](https://github.com/PagerDuty/go-pagerduty/issues/394)

### CODE REVIEW

File: audit.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	NextCursor       string           `json:"next_cursor,omitempty"`
+}
+
+// AuditRecord is a audit trail record that matches the query criteria.
+type AuditRecord struct {
+	ID               string           `json:"id,omitempty"`
+	Self             string           `json:"self,omitempty"`
+	ExecutionTime    string           `json:"execution_time,omitempty"`
+	ExecutionContext ExecutionContext `json:"execution_context,omitempty"`
+	Actors           []Actor          `json:"actors,omitempty"`
```
Comment (theckman) [inline audit.go:43]: Let's switch this to be `[]APIObject`, I'll explain why in my comment on the `Actor` type definition...

File: audit.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+// ExecutionContext contains information about the action execution context.
+type ExecutionContext struct {
+	RequestID     string `json:"request_id,omitempty"`
+	RemoteAddress string `json:"remote_address,omitempty"`
+}
+
+// Actor is the actor who carried out the action.
+type Actor APIObject
```
Comment (theckman) [inline audit.go:62]: I think we should avoid this becoming a named type. If PagerDuty ever adds any functionality in the future, where something else is referred to as an "actor", we can no longer use this name without a breaking change. I think using it directly as an `APIObject` is more clear to consumers that it's not a hydrated object.

File: audit.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+// Resource contains information on the resource that was audited.
+type Resource APIObject
```
Comment (theckman) [inline audit.go:65]: I have the same thoughts here as I do about `Actor`.

File: audit.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+type Reference struct {
+	Name        string      `json:"name,omitempty"`
+	Description string      `json:"description,omitempty"`
+	Added       []APIObject `json:"added,omitempty"`
+	Removed     []APIObject `json:"removed,omitempty"`
+}
+
+// ListAuditRecords lists audit trial records matching provided query params
+// or default criteria.
+func (c *Client) ListAuditRecords(ctx context.Context, o ListAuditRecordsOptions) (*ListAuditRecordsResponse, error) {
```
Comment (theckman) [inline audit.go:103]: I think per issue #270 we should return `(ListAuditRecordsResponse, error)`.

File: audit.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	if err = c.decodeJSON(resp, &result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+
+}
+
+// ListAuditRecordsPaginated lists audit trial records matching provided query
+// params or default criteria, processing paginated responses.
```
Comment (theckman) [inline audit.go:125]: Could you add these to sentences to the end of this comment:

```
The include function decides whether or not to include a specific AuditRecord in the final result. If the include function is nil, all audit records from the API are included by default.
```

The rest looks good!

File: client.go [hunk 6]
```diff
@@ -656,3 +656,53 @@ func (c *Client) pagedGet(ctx context.Context, basePath string, handler response
...
+		if len(c.NextCursor) == 0 {
+			break
+		}
+
+		// Otherwise, update next to parse more results.
+		next = c.NextCursor
+	}
+
+	return nil
+}
```
Comment (theckman) [inline client.go:708]: Thank you for doing all of this! 👍

**Review by theckman — CHANGES_REQUESTED** [review]
Thank you so much! This looks almost ready, with a few small requested changes.

Could you also squash your commits down to a single one just to help keep the history cleaner once it's merged?

**Comment by t-junjie** [comment]
I pulled in changes from `master` and made the relevant changes as requested above. 
It looks like I overwrote some of the commits when I squashed them when I did a `git rebase`, causing the previous review to be "outdated". Sorry about that

**Comment by theckman** [comment]
The tests are failing:

```
=== RUN   TestAudit_List
    audit_test.go:94: values not equal (-want / +got):
          interface{}(
        - 	&pagerduty.ListAuditRecordsResponse{
        - 		Records: []pagerduty.AuditRecord{
        - 			{
        - 				ID:               "PDRECORDID4_UPDATED_USERS_NOTIFI"...,
        - 				ExecutionTime:    "2020-06-04T15:30:16.272Z",
        - 				ExecutionContext: pagerduty.ExecutionContext{...},
        - 				Actors:           []pagerduty.APIObject{...},
        - 				...
        - 			},
        - 		},
        - 		Limit: 10,
        - 	},
        + 	pagerduty.ListAuditRecordsResponse{
        + 		Records: []pagerduty.AuditRecord{
        + 			{
        + 				ID:               "PDRECORDID4_UPDATED_USERS_NOTIFI"...,
        + 				ExecutionTime:    "2020-06-04T15:30:16.272Z",
        + 				ExecutionContext: pagerduty.ExecutionContext{...},
        + 				Actors:           []pagerduty.APIObject{...},
        + 				...
        + 			},
        + 		},
        + 		Limit: 10,
        + 	},
          )
--- FAIL: TestAudit_List (0.00s)
````

**Comment by t-junjie** [comment]
I made the appropriate changes and the test has passed.

**Review by theckman — APPROVED** [review]

**Pull request merged**